core/src/Kokkos_CopyViews.hpp

```diff
@@ -302,7 +302,14 @@ struct ViewCopy<ViewTypeA, ViewTypeB, Layout, ExecSpace, 1, iType> {
   }
 
   KOKKOS_INLINE_FUNCTION
-  void operator()(const iType& i0) const { a(i0) = b(i0); }
+  void operator()(const iType& i0) const {
+    // backwards compatible fix for allowing deep_copy between
+    // non-assignable types for rank 0-3
+    if constexpr (std::is_assignable_v<decltype(a(i0)), decltype(b(i0))>)
+      a(i0) = b(i0);
+    else
+      a(i0) = static_cast<value_type>(b(i0));
+  }
```

Comment:
1. This logic gets repeated 3 times (ranks 1, 2 and 3). I would suggest a macro in this case:
   ```c++
   #define KOKKOS_IMPL_COPY_VIEW_ASSIGNABLE_OR_CAST(dst, src)          \
     if constexpr (std::is_assignable_v<decltype(dst), decltype(src)>) \
       dst = src;                                                      \
     else                                                              \
       dst = static_cast<value_type>(src);
   ``` 
   and your fix becomes
   ```c++
   KOKKOS_IMPL_COPY_VIEW_ASSIGNABLE_OR_CAST(a(i0), b(i0))
   ```
2. Why not fixing higher ranks as well ?

Reply:
I really want to avoid macros as much as possible, and we generally are very unsure that deep_copy should ever do a `static_cast`. This explicitly only enables it for cases where it used to work, without digging a deeper hole.